control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -147,97 +151,113 @@ pub(super) async fn handle_faulted_children(
+    nexus.warn_span(|| {
+        tracing::warn!(%child.uri, %child.state, %child.state_reason, ?faulted_at, "Attempting to remove faulted child")
     });
     nexus
         .remove_child_by_uri(context.registry(), &nexus_state, &child.uri, true)
         .await?;
 
-    nexus_spec_clone.warn_span(|| {
-        tracing::warn!(%child.uri, %child.state, %child.state_reason, "Successfully removed faulted child")
+    nexus.warn_span(|| {
+        tracing::warn!(%child.uri, %child.state, %child.state_reason, ?faulted_at, "Successfully removed faulted child")
     });
     Ok(())
 }
 
-/// Returns true if time elapsed from child fault time is greater or equal to wait time duration.
+/// Returns elapsed time, if time elapsed from child fault time is greater or equal to the
+/// wait time duration.
 fn wait_duration_elapsed(
-    fault_time: Option<SystemTime>,
+    child_uri: &ChildUri,
+    replica_uuid: &ReplicaId,
+    fault_time: SystemTime,
     wait_duration: Duration,
-    uri: &ChildUri,
-) -> bool {
-    if let Some(fault_time) = fault_time {
-        if let Ok(elapsed) = fault_time.elapsed() {
-            info!(
-                child.uri = %uri,
-                "waited {:?} for child to comeback. Wait time is {:?}",
-                elapsed, wait_duration
+) -> Option<Duration> {
+    match fault_time.elapsed() {
+        Ok(elapsed) if elapsed >= wait_duration => Some(elapsed),
+        Ok(elapsed) => {
+            tracing::info!(
+                child.uri = %child_uri,
+                replica.uuid = %replica_uuid,
+                "Waited {elapsed:?} for the child's replica to come back (wait time is {wait_duration:?})"
             );
-            return elapsed >= wait_duration;
+            None
+        }
+        // if time drifts by this much then just keep waiting
+        Err(error) if error.duration() < Duration::from_secs(5) => None,
+        // otherwise just rebuild now anyway, assume wait_duration expired.
+        Err(error) => {
+            tracing::error!(
+                "SystemTime has drifted (time is -{error:?}), assuming wait-duration expired as worst-case"
+            );
+            Some(wait_duration)
         }
     }
-    false
 }
 
-/// If we can get replica for a child from registry. It means that child node is online, Pool is
-/// imported containing the child/replica.
-async fn get_child_replica(
-    nexus: &NexusSpec,
-    child: &Child,
-    context: &PollContext,
-) -> Result<Replica, SvcError> {
-    let replica_uri = nexus.replica_uri(&child.uri).ok_or(SvcError::NotFound {
-        kind: ResourceKind::Replica,
-        id: "Child replica not found".to_string(),
-    })?;
-    context.registry().replica(replica_uri.uuid()).await
+/// Check if a child's replica is online.
+/// If the node has gone offline the replica will not be present in the registry.
+async fn child_replica_online(child_uuid: &ReplicaId, context: &PollContext) -> bool {
```

Comment:
nit: I would call this child_replica_is_online() to flag that it is returning a bool

Reply:
Indeed, done